### PR TITLE
sinkv2(cdc): fix panics about table scheduling or blackhole sink

### LINKED ISSUE #8024: CDC panic: onResolvedTsMsg is called for an invalid table, when running redo log changefeed against GCS
### What did you do?

- deploy tidb cluster
- create eventual consistency changefeed, and use GCS as target storage
- Run workload

### What did you expect to see?

CDC should run without panic

### What did you see instead?

CDC panic

```
kubectl logs -p upstream-ticdc-0 --kubeconfig kubeconfig.yml -n cdc-fb-x7bfs
[WARN] TiCDC server data-dir is not set. Please use `cdc server --data-dir` to start the cdc server if possible.
panic: onResolvedTsMsg is called for an invalid table

goroutine 811 [running]:
[github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).onResolvedTsMsg](http://github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).onResolvedTsMsg)(0xc0021dda00, {0x9c, {0xc0461f4ea0, 0x12, 0x18}, {0xc0461f4eb8, 0x12, 0x18}}, 0x616025cac98000d)
        [github.com/pingcap/tiflow/cdc/redo/manager.go:481](http://github.com/pingcap/tiflow/cdc/redo/manager.go:481) +0x1c5
[github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).bgUpdateLog.func3()](http://github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).bgUpdateLog.func3())
        [github.com/pingcap/tiflow/cdc/redo/manager.go:557](http://github.com/pingcap/tiflow/cdc/redo/manager.go:557) +0x9cb
[github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).bgUpdateLog](http://github.com/pingcap/tiflow/cdc/redo.(*ManagerImpl).bgUpdateLog)(0xc0021dda00, {0x4231230, 0xc002675da0}, 0x7d0, 0xc0021c23c0)
        [github.com/pingcap/tiflow/cdc/redo/manager.go:597](http://github.com/pingcap/tiflow/cdc/redo/manager.go:597) +0x93c
created by [github.com/pingcap/tiflow/cdc/redo.NewManager](http://github.com/pingcap/tiflow/cdc/redo.NewManager)
        [github.com/pingcap/tiflow/cdc/redo/manager.go:187](http://github.com/pingcap/tiflow/cdc/redo/manager.go:187) +0x46b
```

### Versions of the cluster

bash-5.1# /cdc version
Release Version: v6.5.0-master-dirty
Git Commit Hash: 941aa3feff4290c93794e0aa44fc081389dd4488
Git Branch: master
UTC Build Time: 2023-01-05 03:34:56
Go Version: go version go1.19.1 linux/amd64
Failpoint Build: false

### LINKED ISSUE #8142: memory quota can panic with blackhole sink
### What did you do?

Create a changefeed with blackhole sink.

### What did you expect to see?

_No response_

### What did you see instead?

TiCDC can panic:
```
panic: memQuota.release fail

goroutine 623 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x2?, 0x2?, {0x0?, 0x0?, 0xc0c70661c0?})
        go.uber.org/zap@v1.23.0/zapcore/entry.go:198 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc006314270, {0xc0e9148e00, 0x2, 0x2})
        go.uber.org/zap@v1.23.0/zapcore/entry.go:264 +0x3ec
go.uber.org/zap.(*Logger).Panic(0xc00008b400?, {0x386edc0?, 0x414705?}, {0xc0e9148e00, 0x2, 0x2})
        go.uber.org/zap@v1.23.0/logger.go:251 +0x59
github.com/pingcap/log.Panic({0x386edc0?, 0xc0e9148e40?}, {0xc0e9148e00?, 0x13?, 0x61819c6f2f00001?})
        github.com/pingcap/log@v1.1.1-0.20221116035753-734d527bc87c/global.go:54 +0x8b
github.com/pingcap/tiflow/cdc/processor/sinkmanager.(*memQuota).release(0xc02a110700, 0x3140be0?, {0xc0022bcb08?, 0x110?, 0x34e65c0?})
        github.com/pingcap/tiflow/cdc/processor/sinkmanager/mem_quota.go:187 +0x414
github.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).GetTableStats(0xc0df5acb40, 0x5e)
        github.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:819 +0x355
github.com/pingcap/tiflow/cdc/processor.(*processor).GetTableStatus(0xc081a19000, 0x5e)
        github.com/pingcap/tiflow/cdc/processor/processor.go:460 +0x9b
github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*table).getAndUpdateTableState(0xc0f4ca5ae0)
        github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/table.go:56 +0x4d
github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*tableManager).poll(0xc02af5f280, {0x7f650963ec10, 0xc0c7f48288})
        github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/table.go:300 +0x125
github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent.(*agent).Tick(0xc000ee70e0, {0x7f650963ec10, 0xc0c7f48288})
        github.com/pingcap/tiflow/cdc/scheduler/internal/v3/agent/agent.go:180 +0x7e
github.com/pingcap/tiflow/cdc/processor.(*processor).tick(0xc081a19000, {0x4144de0?, 0xc0c7f48288})
        github.com/pingcap/tiflow/cdc/processor/processor.go:716 +0x1b8
```

### Versions of the cluster

master.

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: close #8024 
Issue Number: close #8142 

### What is changed and how it works?

* For #8024, `onResolvedTsMsg` can be called after table is removed, so just remove the panic code.
* For #8142, `memquota.record` is called after `memquota.release` when blackhole sink is used, so correct the order.
* Also add sort engine read metrics.

#### Tests 

 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/redo/manager.go [hunk 1]
```diff
@@ -493,11 +493,10 @@ func (m *ManagerImpl) flushLog(
 }
 
 func (m *ManagerImpl) onResolvedTsMsg(span tablepb.Span, resolvedTs model.Ts) {
-	value, loaded := m.rtsMap.Load(span)
-	if !loaded {
-		panic("onResolvedTsMsg is called for an invalid table")
+	// It's possible that the table is removed while redo log is still in writing.
```
Comment (CharlesCheung96) [inline cdc/redo/manager.go:496]: Why does this happen？

File: cdc/processor/sourcemanager/engine/engine.go [hunk 2]
```diff
@@ -49,13 +49,14 @@ type SortEngine interface {
...
-	FetchByTable(span tablepb.Span, lowerBound, upperBound Position) EventIterator
+	FetchByTable(changefeedID model.ChangeFeedID, span tablepb.Span,
+		lowerBound, upperBound Position) EventIterator
 
 	// FetchAllTables creates an iterator to fetch events from all tables.
 	// lowerBound is inclusive and only resolved events can be retrieved.
 	//
 	// NOTE: It's only available if IsTableBased returns false.
-	FetchAllTables(lowerBound Position) EventIterator
+	FetchAllTables(changefeedID model.ChangeFeedID, lowerBound Position) EventIterator
```
[the remark this answers is not in the document]
Comment (hicqu) [reply cdc/processor/sourcemanager/engine/engine.go:59]: Good idea.

File: cdc/redo/manager.go (hunk 1 above)
Comment (hicqu) [reply cdc/redo/manager.go:496]: `redo_log_worker` is running background. `SinkManager.AsyncStopTable` won't wait all tasks get finished in that worker. So a table can be removed before `onResolvedTsMsg` is called.

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**